python/instrumentation/openinference-instrumentation-dspy/src/openinference/instrumentation/dspy/__init__.py

```diff
@@ -0,0 +1,169 @@
+import json
+from abc import ABC
+from typing import Any, Callable, Collection, Dict, Mapping, Tuple
+
+from openinference.instrumentation.dspy.package import _instruments
+from openinference.instrumentation.dspy.version import __version__
+from openinference.semconv.trace import (
+    OpenInferenceMimeTypeValues,
+    OpenInferenceSpanKindValues,
+    SpanAttributes,
+)
+from opentelemetry import trace as trace_api
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor  # type: ignore
+from wrapt import wrap_function_wrapper
+
+_DSPY_MODULE = "dspy"
+
+# DSPy used to be called DSP - some of the modules still fall under the old namespace
+_DSP_MODULE = "dsp"
+
+
+class DSPyInstrumentor(BaseInstrumentor):  # type: ignore
+    """
+    OpenInference Instrumentor for DSPy
+    """
+
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return _instruments
+
+    def _instrument(self, **kwargs: Any) -> None:
+        if not (tracer_provider := kwargs.get("tracer_provider")):
+            tracer_provider = trace_api.get_tracer_provider()
+        tracer = trace_api.get_tracer(__name__, __version__, tracer_provider)
+
+        # Instrument LM (language model) calls
+        from dsp.modules.lm import LM
+
+        language_model_classes = LM.__subclasses__()
+        for lm in language_model_classes:
+            wrap_function_wrapper(
+                module=_DSP_MODULE,
+                name=lm.__name__ + ".basic_request",
+                wrapper=_LMBasicRequestWrapper(tracer),
+            )
+
+        # Instrument DSPy constructs
+        wrap_function_wrapper(
+            module=_DSPY_MODULE,
+            name="Predict.forward",
+            wrapper=_PredictForwardWrapper(tracer),
+        )
+
+    def _uninstrument(self, **kwargs: Any) -> None:
+        from dsp.modules.lm import LM
+
+        language_model_classes = LM.__subclasses__()
+        for lm in language_model_classes:
+            if hasattr(lm.request, "__wrapped__"):
+                lm.request = lm.request.__wrapped__
+
+        # Restore DSPy constructs
+        from dspy import Predict
+
+        if hasattr(Predict.forward, "__wrapped__"):
+            Predict.forward = Predict.forward.__wrapped__
+
+
+class _WithTracer(ABC):
+    """
+    Base class for wrappers that need a tracer. Acts as a trait for the wrappers
+    """
+
+    def __init__(self, tracer: trace_api.Tracer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._tracer = tracer
+
+
+class _LMBasicRequestWrapper(_WithTracer):
+    """
+    Wrapper for DSP LM.basic_request
+    Captures all calls to language models (lm)
+    """
+
+    def __call__(
+        self,
+        wrapped: Callable[..., Any],
+        instance: Any,
+        args: Tuple[type, Any],
```

Comment:
this type signature needs to be updated

Reply:
I'm not sure I follow what it should be - lmk. Gonna merge and will fix in a follow-up